pyhf/pdf.py

```diff
@@ -181,36 +220,33 @@ def expected_sample(self, channel, sample, pars):
         # \phi == normfactor, overallsys
         # \sigma == histosysdelta + nominal
 
-        tensorlib, _ = get_backend()
         # first, collect the factors from all modifiers
-        results = {'shapesys': [],
-                   'normfactor': [],
-                   'shapefactor': [],
-                   'staterror': [],
-                   'histosys': [],
-                   'normsys': []}
-        for m in sample['modifiers']:
-            modifier, modpars = self.config.modifier(m['name']), pars[self.config.par_slice(m['name'])]
-            results[m['type']].append(modifier.apply(channel, sample, modpars))
-
-
-        # start building the entire set of factors
-        factors = []
-        # scalars that get broadcasted to shape of vectors
-        factors += results['normfactor']
-        factors += results['normsys']
-        # vectors
-        factors += results['shapesys']
-        factors += results['shapefactor']
-        factors += results['staterror']
-
-        nominal = tensorlib.astensor(sample['data'])
-        factors += [tensorlib.sum(tensorlib.stack([
-          nominal,
-          tensorlib.sum(tensorlib.stack(results['histosys']), axis=0)
-        ]), axis=0) if len(results['histosys']) > 0 else nominal]
 
-        return tensorlib.product(tensorlib.stack(tensorlib.simple_broadcast(*factors)), axis=0)
+        all_modifications = {}
+        factor_mods = ['normfactor','normsys','shapesys','shapefactor','staterror']
+        delta_mods  = ['histosys']
+
+        all_results = {}
+        for mtype in factor_mods + delta_mods:
+            all_results[mtype] = self._mtype_results(mtype,pars)
+
+        for channel in self.spec['channels']:
+            for sample in channel['samples']:
+                #sum of lists is concat
+                factors = sum([
+                    all_results.get(x,{}).get(channel['name'],{}).get(sample['name'],[])
+                    for x in factor_mods
+                ],[]) 
+                deltas  = sum([
+                    all_results.get(x,{}).get(channel['name'],{}).get(sample['name'],[])
+                    for x in delta_mods
+                ],[])
+                
+                all_modifications.setdefault(
+                    channel['name'],{})[
+                    sample['name']
+                ] = (factors, deltas)
+        return all_modifications
```

Comment:
this function implements the calculation of *all* modifications by looping over all possible modifications and calling `_mtype_results ` for each one. We then acumulate the results in a nested dict-like structure to keep track of factors/deltas that is then used by `expected_actualdata`/`expected_sample`

Reply:
the names / roles of the modifiers are hardcoded for now, if we can get more expressive labels / attributes on the modifiers that signify what type of modifier they are (factor / delta) this hard-coding could be removed

Reply:
Not hard-coded anymore ;)